{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Insight Generation Pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import all the important stuff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### General Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import getpass\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Langchain Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain import hub\n",
    "from langchain.chat_models import init_chat_model\n",
    "from langchain_community.utilities import SQLDatabase\n",
    "from langchain_community.tools.sql_database.tool import QuerySQLDatabaseTool\n",
    "from langchain_core.messages import AIMessage, HumanMessage\n",
    "from langchain_core.prompts import ChatPromptTemplate, PromptTemplate, FewShotPromptTemplate\n",
    "from langchain_ollama import ChatOllama\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### OpenAI Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from openai import OpenAI"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LLM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain.chat_models import init_chat_model\n",
    "if not os.environ.get(\"OPENAI_API_KEY\"):\n",
    "  os.environ[\"OPENAI_API_KEY\"] = getpass.getpass(\"Enter API key for OpenAI: \")\n",
    "model = init_chat_model(\"gpt-4o-mini\", model_provider=\"openai\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SQL agent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_community.agent_toolkits import create_sql_agent\n",
    "from langchain_core.prompts import FewShotPromptTemplate, PromptTemplate, ChatPromptTemplate\n",
    "from langchain.agents import AgentExecutor, Tool, AgentType, initialize_agent\n",
    "from langchain.agents.openai_functions_agent.base import OpenAIFunctionsAgent\n",
    "from langchain.agents.agent import AgentExecutor\n",
    "import json\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "demonstration_sqls_15B_3S_oncomx.json\n"
     ]
    }
   ],
   "source": [
    "data = []\n",
    "file_names = [\n",
    "    \"demonstration_sqls_15B_3S_oncomx.json\",\n",
    "    ]\n",
    "directory = \"dem_sqls_oncomx\"\n",
    "\n",
    "for file_name in file_names:\n",
    "    print(file_name)\n",
    "    with open(os.path.join(directory, file_name), \"r\") as f:\n",
    "        data.append(json.load(f))\n",
    "\n",
    "#demonstration_sqls = json.load(open(\"dem_sqls/demonstration_sqls_cordis_1B_1S.json\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CREATE TABLE anatomical_entity (\n",
      "\tid VARCHAR NOT NULL, \n",
      "\tname VARCHAR, \n",
      "\tdescription TEXT, \n",
      "\tCONSTRAINT idx_18385_primary PRIMARY KEY (id)\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from anatomical_entity table:\n",
      "id\tname\tdescription\t\n",
      "CL:0000057\tfibroblast\tA connective tissue cell which secretes an extracellular matrix rich in collagen and other macromolecules. Flattened and irregular in outline with branching processes; appear fusiform or spindle-shaped.\t\n",
      "CL:0000169\ttype B pancreatic cell\tA cell that secretes insulin and is located towards the center of the islets of Langerhans.\t\n",
      "CL:0002092\tbone marrow cell\tA cell found in the bone marrow. This can include fibroblasts, macrophages, adipocytes, osteoblasts, osteoclasts, endothelial cells and hematopoietic cells.\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE biomarker (\n",
      "\tid VARCHAR DEFAULT '0', \n",
      "\tgene_symbol VARCHAR, \n",
      "\tbiomarker_description TEXT, \n",
      "\tbiomarker_id VARCHAR, \n",
      "\ttest_is_a_panel INTEGER NOT NULL, \n",
      "\tCONSTRAINT idx_18391_primary PRIMARY KEY (id)\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from biomarker table:\n",
      "id\tgene_symbol\tbiomarker_description\tbiomarker_id\ttest_is_a_panel\t\n",
      "EDRN00003\tICAM1\tFrom NCBI Gene: This gene encodes a cell surface glycoprotein which is typically expressed on endothelial cells and cells of the immune system. It binds to integrins of type CD11a / CD18, or CD11b / CD18 and is also exploited by Rhinovirus as a receptor. [provided by RefSeq, Jul 2008]\turn:edrn:bmdb:biomarker:101\tf\t\n",
      "EDRN00004\tADIPOQ\tFrom NCBI Gene: This gene is expressed in adipose tissue exclusively. It encodes a protein with similarity to collagens X and VIII and complement factor C1q. The encoded protein circulates in the plasma and is involved with metabolic and hormonal processes. Mutations in this gene are associated with adiponectin deficiency. Multiple alternatively spliced variants, encoding the same protein, have been identified. [provided by RefSeq, Apr 2010]\turn:edrn:bmdb:biomarker:1016\tf\t\n",
      "EDRN00005\tAKR1B1\tFrom NCBI Gene: This gene encodes a member of the aldo/keto reductase superfamily, which consists of more than 40 known enzymes and proteins. This member catalyzes the reduction of a number of aldehydes, including the aldehyde form of glucose, and is thereby implicated in the development of diabetic complications by catalyzing the reduction of glucose to sorbitol. Multiple pseudogenes have been identified for this gene. The nomenclature system used by the HUGO Gene Nomenclature Committee to define human aldo-keto reductase family members is known to differ from that used by the Mouse Genome Informatics database. [provided by RefSeq, Feb 2009]\turn:edrn:bmdb:biomarker:1017\tf\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE disease (\n",
      "\tid INTEGER NOT NULL, \n",
      "\tname VARCHAR NOT NULL, \n",
      "\tCONSTRAINT idx_18463_primary PRIMARY KEY (id)\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from disease table:\n",
      "id\tname\t\n",
      "0\tundefined\t\n",
      "119\tvaginal cancer\t\n",
      "169\tneuroendocrine tumor\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE map_uniprot_canonical_id (\n",
      "\tuniprotkb_ac VARCHAR NOT NULL, \n",
      "\tuniprotkb_canonical_ac VARCHAR NOT NULL, \n",
      "\tCONSTRAINT idx_18520_primary PRIMARY KEY (uniprotkb_ac)\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from map_uniprot_canonical_id table:\n",
      "uniprotkb_ac\tuniprotkb_canonical_ac\t\n",
      "A0A075B6F4\tA0A075B6F4-1\t\n",
      "A0A075B6H5\tA0A075B6H5-1\t\n",
      "A0A075B6H7\tA0A075B6H7-1\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE species (\n",
      "\tspeciesid INTEGER NOT NULL, \n",
      "\tgenus VARCHAR NOT NULL, \n",
      "\tspecies VARCHAR NOT NULL, \n",
      "\tspeciescommonname VARCHAR NOT NULL, \n",
      "\tCONSTRAINT idx_18526_primary PRIMARY KEY (speciesid)\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from species table:\n",
      "speciesid\tgenus\tspecies\tspeciescommonname\t\n",
      "9606\tHomo\tsapiens\thuman\t\n",
      "10090\tMus\tmusculus\tmouse\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE stage (\n",
      "\tid VARCHAR NOT NULL, \n",
      "\tname VARCHAR NOT NULL, \n",
      "\tCONSTRAINT idx_18529_primary PRIMARY KEY (id)\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from stage table:\n",
      "id\tname\t\n",
      "HsapDv:0000084\t2-5 year-old child stage (human)\t\n",
      "HsapDv:0000085\t6-12 year-old child stage (human)\t\n",
      "HsapDv:0000086\tadolescent stage (human)\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE biomarker_alias (\n",
      "\tbiomarker_internal_id VARCHAR NOT NULL, \n",
      "\talias VARCHAR, \n",
      "\tCONSTRAINT idx_18398_primary PRIMARY KEY (biomarker_internal_id, alias), \n",
      "\tCONSTRAINT fk_biomarker_alias FOREIGN KEY(biomarker_internal_id) REFERENCES biomarker (id) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from biomarker_alias table:\n",
      "biomarker_internal_id\talias\t\n",
      "EDRN00261\t\tCCCTC-binding factor (zinc finger protein)-like\t\n",
      "EDRN00244\t\tdickkopf homolog 3 (Xenopus laevis)\t\n",
      "EDRN00586\t\tEGF containing fibulin-like extracellular matrix protein 1\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE biomarker_article (\n",
      "\tbiomarker_internal_id VARCHAR DEFAULT '0', \n",
      "\tpmid VARCHAR, \n",
      "\tCONSTRAINT idx_18401_primary PRIMARY KEY (biomarker_internal_id, pmid), \n",
      "\tCONSTRAINT fk_biomarker_article FOREIGN KEY(biomarker_internal_id) REFERENCES biomarker (id) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from biomarker_article table:\n",
      "biomarker_internal_id\tpmid\t\n",
      "EDRN00403\t10101297\t\n",
      "EDRN00506\t10371572\t\n",
      "EDRN00402\t10397259\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE biomarker_edrn (\n",
      "\tid VARCHAR DEFAULT '0', \n",
      "\tqa_state VARCHAR NOT NULL, \n",
      "\tbiomarker_title VARCHAR NOT NULL, \n",
      "\tbiomarker_type VARCHAR NOT NULL, \n",
      "\tuberon_anatomical_id VARCHAR NOT NULL, \n",
      "\tphase VARCHAR, \n",
      "\tCONSTRAINT idx_18405_primary PRIMARY KEY (id), \n",
      "\tCONSTRAINT fk_biomarker_edrn_anat FOREIGN KEY(uberon_anatomical_id) REFERENCES anatomical_entity (id) ON DELETE CASCADE ON UPDATE CASCADE, \n",
      "\tCONSTRAINT fk_biomarker_edrn_id FOREIGN KEY(id) REFERENCES biomarker (id) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from biomarker_edrn table:\n",
      "id\tqa_state\tbiomarker_title\tbiomarker_type\tuberon_anatomical_id\tphase\t\n",
      "EDRN00001\tUnder Review\tMicrosatellite Instability Assay\tGenetic\tUBERON:0018707\t\t\n",
      "EDRN00002\tUnder Review\tGarcia chromosomal aneusomy + cytology\tProtein\tUBERON:0018707\t\t\n",
      "EDRN00003\tCurated\tICAM1\tProtein\tUBERON:0000310\t\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE biomarker_fda_test (\n",
      "\ttest_trade_name VARCHAR NOT NULL, \n",
      "\ttest_manufacturer VARCHAR NOT NULL, \n",
      "\ttest_submission VARCHAR NOT NULL, \n",
      "\tbiomarker_origin VARCHAR NOT NULL, \n",
      "\tdoid INTEGER NOT NULL, \n",
      "\thistological_type VARCHAR NOT NULL, \n",
      "\tspecimen_type VARCHAR NOT NULL, \n",
      "\tplatform_method VARCHAR NOT NULL, \n",
      "\ttest_number_genes INTEGER NOT NULL, \n",
      "\ttest_adoption_evidence VARCHAR NOT NULL, \n",
      "\ttest_approval_status VARCHAR NOT NULL, \n",
      "\ttest_study_design VARCHAR, \n",
      "\tclinical_significance TEXT NOT NULL, \n",
      "\tCONSTRAINT idx_18433_primary PRIMARY KEY (test_submission, test_trade_name), \n",
      "\tCONSTRAINT fk_biomarker_fda_test_doid FOREIGN KEY(doid) REFERENCES disease (id) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from biomarker_fda_test table:\n",
      "test_trade_name\ttest_manufacturer\ttest_submission\tbiomarker_origin\tdoid\thistological_type\tspecimen_type\tplatform_method\ttest_number_genes\ttest_adoption_evidence\ttest_approval_status\ttest_study_design\tclinical_significance\t\n",
      "VENTANA ANTI-BRAF V600E (VE1) MOUSE MONOCLONAL PRIMARY ANTIBODY\tVentana Medical Systems\tDEN170030\tmultigene prediction\t9256\tC209 Tumors may be present in a single or multiple segments of the colon, rectosigmoid, rectum.\tparrafin block\tICH\t5\tclinical use\t513(f)(2)\tRetObs\tThe VENTANA MMR IHC Panel is a qualitative immunohistochemistry (IHC) test intended for use in the light microscopic assessment of mismatch repair (MMR) proteins (MLH1, PMS2, MSH2, and MSH6) and BRAF V600E proteins in formalin-fixed, paraffin-embedded colorectal cancer (CRC) tissue sections.\t\n",
      "VENTANA ANTI-MLH-1(M1) MOUSE MONOCLONAL PRIMARY ANTIBODY\tVentana Medical Systems\tDEN170030\tmultigene prediction\t9256\tC209 Tumors may be present in a single or multiple segments of the colon, rectosigmoid, rectum.\tparrafin block\tICH\t5\tclinical use\t513(f)(2)\tRetObs\tThe VENTANA MMR IHC Panel is a qualitative immunohistochemistry (IHC) test intended for use in the light microscopic assessment of mismatch repair (MMR) proteins (MLH1, PMS2, MSH2, and MSH6) and BRAF V600E proteins in formalin-fixed, paraffin-embedded colorectal cancer (CRC) tissue sections.\t\n",
      "VENTANA ANTI-MSH2 (G219-1129) MOUSE MONOCLONAL PRIMARY ANTIBODY\tVentana Medical Systems\tDEN170030\tmultigene prediction\t9256\tC209 Tumors may be present in a single or multiple segments of the colon, rectosigmoid, rectum.\tparrafin block\tICH\t5\tclinical use\t513(f)(2)\tRetObs\tThe VENTANA MMR IHC Panel is a qualitative immunohistochemistry (IHC) test intended for use in the light microscopic assessment of mismatch repair (MMR) proteins (MLH1, PMS2, MSH2, and MSH6) and BRAF V600E proteins in formalin-fixed, paraffin-embedded colorectal cancer (CRC) tissue sections.\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE cancer_tissue (\n",
      "\tdoid INTEGER NOT NULL, \n",
      "\tuberon_anatomical_id VARCHAR NOT NULL, \n",
      "\tCONSTRAINT idx_26577_primary PRIMARY KEY (doid, uberon_anatomical_id), \n",
      "\tCONSTRAINT fk_doid_cancer_tissue FOREIGN KEY(doid) REFERENCES disease (id) ON DELETE CASCADE ON UPDATE CASCADE, \n",
      "\tCONSTRAINT fk_uberon_cancer_tissue FOREIGN KEY(uberon_anatomical_id) REFERENCES anatomical_entity (id) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from cancer_tissue table:\n",
      "doid\tuberon_anatomical_id\t\n",
      "219\tUBERON:0001155\t\n",
      "263\tUBERON:0002113\t\n",
      "363\tUBERON:0003134\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE differential_expression (\n",
      "\tgene_symbol VARCHAR NOT NULL, \n",
      "\tdoid INTEGER NOT NULL, \n",
      "\tlog2fc FLOAT NOT NULL, \n",
      "\tpvalue FLOAT NOT NULL, \n",
      "\tadjpvalue FLOAT NOT NULL, \n",
      "\tstatistical_significance VARCHAR NOT NULL, \n",
      "\texpression_change_direction VARCHAR NOT NULL, \n",
      "\tsubjects_up INTEGER NOT NULL, \n",
      "\tsubjects_down INTEGER NOT NULL, \n",
      "\tsubjects_nochange INTEGER NOT NULL, \n",
      "\tsubjects_nocoverage INTEGER NOT NULL, \n",
      "\tsubjects_total INTEGER NOT NULL, \n",
      "\tCONSTRAINT idx_26580_primary PRIMARY KEY (gene_symbol, doid), \n",
      "\tCONSTRAINT fk_df_doid FOREIGN KEY(doid) REFERENCES disease (id) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from differential_expression table:\n",
      "gene_symbol\tdoid\tlog2fc\tpvalue\tadjpvalue\tstatistical_significance\texpression_change_direction\tsubjects_up\tsubjects_down\tsubjects_nochange\tsubjects_nocoverage\tsubjects_total\t\n",
      "A1BG\t263\t0.91\t1.57e-14\t6.31e-14\tYes\tup\t30\t33\t19\t0\t82\t\n",
      "A1BG\t363\t-0.52\t0.014\t0.028\tYes\tdown\t4\t10\t2\t0\t16\t\n",
      "A1BG\t1324\t0.22\t0.0571\t0.0798\tNo\tup\t18\t33\t27\t0\t78\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE map_protein_disease_mutation (\n",
      "\tpeptide_id VARCHAR NOT NULL, \n",
      "\tensembl_transcript_id VARCHAR NOT NULL, \n",
      "\tuniprotkb_ac VARCHAR NOT NULL, \n",
      "\tCONSTRAINT idx_18517_primary PRIMARY KEY (ensembl_transcript_id), \n",
      "\tCONSTRAINT fk_uniprotkb_ac FOREIGN KEY(uniprotkb_ac) REFERENCES map_uniprot_canonical_id (uniprotkb_ac) ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from map_protein_disease_mutation table:\n",
      "peptide_id\tensembl_transcript_id\tuniprotkb_ac\t\n",
      "ENSP00000000233\tENST00000000233\tP84085\t\n",
      "ENSP00000000412\tENST00000000412\tP20645\t\n",
      "ENSP00000000442\tENST00000000442\tP11474\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE xref_gene_ensembl (\n",
      "\tgene_symbol VARCHAR NOT NULL, \n",
      "\tensembl_gene_id VARCHAR NOT NULL, \n",
      "\tspeciesid INTEGER NOT NULL, \n",
      "\tCONSTRAINT idx_18532_primary PRIMARY KEY (ensembl_gene_id), \n",
      "\tCONSTRAINT fk_gene_speciesid FOREIGN KEY(speciesid) REFERENCES species (speciesid) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from xref_gene_ensembl table:\n",
      "gene_symbol\tensembl_gene_id\tspeciesid\t\n",
      "TSPAN6\tENSG00000000003\t9606\t\n",
      "TNMD\tENSG00000000005\t9606\t\n",
      "DPM1\tENSG00000000419\t9606\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE biomarker_fda (\n",
      "\tid VARCHAR DEFAULT '0', \n",
      "\ttest_trade_name VARCHAR NOT NULL, \n",
      "\ttest_submission VARCHAR NOT NULL, \n",
      "\tCONSTRAINT idx_18415_primary PRIMARY KEY (id), \n",
      "\tCONSTRAINT fk_biomarker_fda_id FOREIGN KEY(id) REFERENCES biomarker (id) ON DELETE CASCADE ON UPDATE CASCADE, \n",
      "\tCONSTRAINT fk_biomarker_fda_test FOREIGN KEY(test_submission, test_trade_name) REFERENCES biomarker_fda_test (test_submission, test_trade_name) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from biomarker_fda table:\n",
      "id\ttest_trade_name\ttest_submission\t\n",
      "FDA000027\tVENTANA ANTI-BRAF V600E (VE1) MOUSE MONOCLONAL PRIMARY ANTIBODY\tDEN170030\t\n",
      "FDA000213\tVENTANA ANTI-BRAF V600E (VE1) MOUSE MONOCLONAL PRIMARY ANTIBODY\tDEN170030\t\n",
      "FDA000229\tVENTANA ANTI-BRAF V600E (VE1) MOUSE MONOCLONAL PRIMARY ANTIBODY\tDEN170030\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE biomarker_fda_test_trial (\n",
      "\ttest_trade_name VARCHAR NOT NULL, \n",
      "\ttest_submission VARCHAR NOT NULL, \n",
      "\ttest_trial_id VARCHAR NOT NULL, \n",
      "\tCONSTRAINT idx_18439_primary PRIMARY KEY (test_trade_name, test_submission, test_trial_id), \n",
      "\tCONSTRAINT fk_biomarker_fda_test_trial FOREIGN KEY(test_submission, test_trade_name) REFERENCES biomarker_fda_test (test_submission, test_trade_name) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from biomarker_fda_test_trial table:\n",
      "test_trade_name\ttest_submission\ttest_trial_id\t\n",
      "23andMe PGS Genetic Health Risk Report for BRCA1/BRCA2 (Selected Variants)\tDEN170046\tNCT01598597\t\n",
      "MammaPrint\tK062694\tNCT00433589\t\n",
      "MammaPrint\tK062695\tNCT00433589\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE biomarker_fda_test_use (\n",
      "\tid INTEGER  NOT NULL, \n",
      "\ttest_trade_name VARCHAR NOT NULL, \n",
      "\ttest_submission VARCHAR NOT NULL, \n",
      "\tapproved_indication VARCHAR NOT NULL, \n",
      "\tactual_use VARCHAR NOT NULL, \n",
      "\tCONSTRAINT idx_18447_primary PRIMARY KEY (id), \n",
      "\tCONSTRAINT fk_biomarker_fda_test_use FOREIGN KEY(test_submission, test_trade_name) REFERENCES biomarker_fda_test (test_submission, test_trade_name) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from biomarker_fda_test_use table:\n",
      "id\ttest_trade_name\ttest_submission\tapproved_indication\tactual_use\t\n",
      "1\tCologuard\tP130017\tdiagnosis\tdiagnostic\t\n",
      "2\tCologuard\tP130017\tdiagnosis\tprognostic\t\n",
      "3\tProsigna Breast Cancer Prognostic Gene Signature Assay\tK130010\tprognosis\tprognostic\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE disease_mutation (\n",
      "\tid INTEGER  NOT NULL, \n",
      "\tchromosome_id VARCHAR NOT NULL, \n",
      "\tchromosome_pos INTEGER NOT NULL, \n",
      "\tref_nt VARCHAR NOT NULL, \n",
      "\talt_nt VARCHAR NOT NULL, \n",
      "\tensembl_transcript_id VARCHAR NOT NULL, \n",
      "\tcds_pos INTEGER NOT NULL, \n",
      "\tpeptide_pos INTEGER NOT NULL, \n",
      "\taa_pos_uniprotkb INTEGER NOT NULL, \n",
      "\tref_aa VARCHAR NOT NULL, \n",
      "\talt_aa VARCHAR NOT NULL, \n",
      "\tmutation_freq INTEGER NOT NULL, \n",
      "\tdata_source VARCHAR NOT NULL, \n",
      "\tdoid INTEGER NOT NULL, \n",
      "\tCONSTRAINT idx_18468_primary PRIMARY KEY (id), \n",
      "\tCONSTRAINT fk_disease_mutation_doid FOREIGN KEY(doid) REFERENCES disease (id) ON DELETE CASCADE ON UPDATE CASCADE, \n",
      "\tCONSTRAINT fk_ensembl_transcript FOREIGN KEY(ensembl_transcript_id) REFERENCES map_protein_disease_mutation (ensembl_transcript_id) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from disease_mutation table:\n",
      "id\tchromosome_id\tchromosome_pos\tref_nt\talt_nt\tensembl_transcript_id\tcds_pos\tpeptide_pos\taa_pos_uniprotkb\tref_aa\talt_aa\tmutation_freq\tdata_source\tdoid\t\n",
      "45\t2\t89278459\tG\tA\tENST00000390247\t45\t15\t15\tL\tL\t2\ttcga\t1909\t\n",
      "51\t2\t89278274\tC\tA\tENST00000390247\t61\t21\t21\tE\t*\t1\ttcga\t363\t\n",
      "71\t2\t89278120\tG\tA\tENST00000390247\t215\t72\t72\tA\tV\t1\ticgc\t3571\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE healthy_expression (\n",
      "\tensembl_gene_id VARCHAR NOT NULL, \n",
      "\tuberon_anatomical_id VARCHAR NOT NULL, \n",
      "\tuberon_developmental_id VARCHAR NOT NULL, \n",
      "\texpression_level_gene_relative VARCHAR NOT NULL, \n",
      "\texpression_level_anatomical_relative VARCHAR NOT NULL, \n",
      "\tcall_quality VARCHAR NOT NULL, \n",
      "\texpression_rank_score FLOAT NOT NULL, \n",
      "\texpression_score FLOAT NOT NULL, \n",
      "\tCONSTRAINT idx_18502_primary PRIMARY KEY (ensembl_gene_id, uberon_anatomical_id, uberon_developmental_id), \n",
      "\tCONSTRAINT fk_ensg_he FOREIGN KEY(ensembl_gene_id) REFERENCES xref_gene_ensembl (ensembl_gene_id) ON DELETE CASCADE ON UPDATE CASCADE, \n",
      "\tCONSTRAINT fk_stage FOREIGN KEY(uberon_developmental_id) REFERENCES stage (id) ON DELETE CASCADE ON UPDATE CASCADE, \n",
      "\tCONSTRAINT fk_uberon_he FOREIGN KEY(uberon_anatomical_id) REFERENCES anatomical_entity (id) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from healthy_expression table:\n",
      "ensembl_gene_id\tuberon_anatomical_id\tuberon_developmental_id\texpression_level_gene_relative\texpression_level_anatomical_relative\tcall_quality\texpression_rank_score\texpression_score\t\n",
      "ENSG00000000003\tUBERON:0001119\tHsapDv:0000088\tHIGH\tHIGH\tSILVER\t6680.0\t87.7\t\n",
      "ENSG00000000003\tUBERON:0001235\tHsapDv:0000090\tHIGH\tHIGH\tSILVER\t9030.0\t83.3\t\n",
      "ENSG00000000003\tUBERON:0001295\tHsapDv:0000092\tHIGH\tHIGH\tGOLD\t4990.0\t90.8\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE biomarker_fda_drug (\n",
      "\tbiomarker_fda_id VARCHAR DEFAULT '0', \n",
      "\tbiomarker_drug VARCHAR, \n",
      "\tCONSTRAINT idx_18419_primary PRIMARY KEY (biomarker_fda_id, biomarker_drug), \n",
      "\tCONSTRAINT fk_biomarker_fda_drug FOREIGN KEY(biomarker_fda_id) REFERENCES biomarker_fda (id) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from biomarker_fda_drug table:\n",
      "biomarker_fda_id\tbiomarker_drug\t\n",
      "FDA000009\tXalkori (crizotinib)\t\n",
      "FDA000032\tVemurafenib\t\n",
      "FDA000033\tVemurafenib\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE biomarker_fda_ncit_term (\n",
      "\tbiomarker_fda_id VARCHAR DEFAULT '0', \n",
      "\tncit_biomarker VARCHAR NOT NULL, \n",
      "\tCONSTRAINT idx_18423_primary PRIMARY KEY (biomarker_fda_id, ncit_biomarker), \n",
      "\tCONSTRAINT fk_biomarker_fda_ncit FOREIGN KEY(biomarker_fda_id) REFERENCES biomarker_fda (id) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from biomarker_fda_ncit_term table:\n",
      "biomarker_fda_id\tncit_biomarker\t\n",
      "FDA000001\tACTB\t\n",
      "FDA000001\tACTB Gene\t\n",
      "FDA000001\tActin\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE disease_mutation_article (\n",
      "\tpmid INTEGER NOT NULL, \n",
      "\tdisease_mutation_id INTEGER NOT NULL, \n",
      "\tCONSTRAINT idx_18472_primary PRIMARY KEY (disease_mutation_id, pmid), \n",
      "\tCONSTRAINT fk_dm_article_id FOREIGN KEY(disease_mutation_id) REFERENCES disease_mutation (id) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from disease_mutation_article table:\n",
      "pmid\tdisease_mutation_id\t\n",
      "1317264\t1637716\t\n",
      "1317264\t1637717\t\n",
      "1317264\t1637720\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE disease_mutation_impact_prediction (\n",
      "\tid INTEGER  NOT NULL, \n",
      "\tdisease_mutation_id INTEGER NOT NULL, \n",
      "\tsite_prediction VARCHAR NOT NULL, \n",
      "\tprobability FLOAT, \n",
      "\ttool VARCHAR NOT NULL, \n",
      "\tCONSTRAINT idx_18477_primary PRIMARY KEY (id), \n",
      "\tCONSTRAINT fk_dm_impact_prediction_id FOREIGN KEY(disease_mutation_id) REFERENCES disease_mutation (id) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from disease_mutation_impact_prediction table:\n",
      "id\tdisease_mutation_id\tsite_prediction\tprobability\ttool\t\n",
      "3\t8778\tpossibly damaging\t0.913\tpolyphen\t\n",
      "11\t8792\tpossibly damaging\t0.913\tpolyphen\t\n",
      "28\t16749\tprobably damaging\t0.99\tpolyphen\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE disease_mutation_site_annotation (\n",
      "\tid INTEGER  NOT NULL, \n",
      "\tdisease_mutation_id INTEGER NOT NULL, \n",
      "\tdescription VARCHAR, \n",
      "\tfeature_key VARCHAR NOT NULL, \n",
      "\tCONSTRAINT idx_18492_primary PRIMARY KEY (id), \n",
      "\tCONSTRAINT fk_dm_site_annotation_id FOREIGN KEY(disease_mutation_id) REFERENCES disease_mutation (id) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from disease_mutation_site_annotation table:\n",
      "id\tdisease_mutation_id\tdescription\tfeature_key\t\n",
      "1\t292\t\tdisulfide bond\t\n",
      "4\t456\t\tdisulfide bond\t\n",
      "5\t577\t\tdisulfide bond\t\n",
      "*/\n",
      "\n",
      "\n",
      "CREATE TABLE disease_mutation_tissue (\n",
      "\tuberon_anatomical_id VARCHAR NOT NULL, \n",
      "\tdisease_mutation_id INTEGER NOT NULL, \n",
      "\tCONSTRAINT idx_18499_primary PRIMARY KEY (uberon_anatomical_id, disease_mutation_id), \n",
      "\tCONSTRAINT idx_fk_disease_mutation_id FOREIGN KEY(disease_mutation_id) REFERENCES disease_mutation (id) ON DELETE CASCADE ON UPDATE CASCADE, \n",
      "\tCONSTRAINT idx_fk_disease_mutation_uberon FOREIGN KEY(uberon_anatomical_id) REFERENCES anatomical_entity (id) ON DELETE CASCADE ON UPDATE CASCADE\n",
      ")\n",
      "\n",
      "/*\n",
      "3 rows from disease_mutation_tissue table:\n",
      "uberon_anatomical_id\tdisease_mutation_id\t\n",
      "UBERON:0000970\t45\t\n",
      "UBERON:0003134\t51\t\n",
      "UBERON:0002107\t71\t\n",
      "*/\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sqlite3\n",
    "def get_database_prompt(db_path) -> str:\n",
    "\n",
    "    stmt = ''\n",
    "\n",
    "    conn = sqlite3.connect(db_path)\n",
    "    cur = conn.cursor()\n",
    "\n",
    "    # Fetch names of all tables\n",
    "    cur.execute(\"SELECT name FROM sqlite_master WHERE type='table';\")\n",
    "    tables = cur.fetchall()\n",
    "\n",
    "    # Fech create statements for all tables\n",
    "    for table in tables:\n",
    "        table_name = table[0]\n",
    "        cur.execute(f\"SELECT sql FROM sqlite_master WHERE type='table' AND name='{table_name}';\")\n",
    "        create_statement = cur.fetchone()[0]\n",
    "\n",
    "        stmt += create_statement + '\\n\\n'\n",
    "        stmt += \"/*\\n\" + f\"3 rows from {table_name} table:\\n\"\n",
    "        \n",
    "        cur.execute(f\"PRAGMA table_info({table_name})\")\n",
    "        columns = cur.fetchall()\n",
    "        for column in columns:\n",
    "            stmt += column[1] + '\\t'\n",
    "        stmt += \"\\n\"\n",
    "            \n",
    "        cur.execute(f\"SELECT * FROM {table_name} LIMIT 3;\")\n",
    "        create_statement = cur.fetchmany(3)\n",
    "        for row in create_statement:\n",
    "            for word in row:\n",
    "                stmt += str(word) + '\\t'\n",
    "            stmt += \"\\n\"\n",
    "        stmt += \"*/\\n\\n\\n\"\n",
    "\n",
    "    conn.close()\n",
    "    return stmt\n",
    "\n",
    "db_path = \"./data/sft_data_collections/oncomx/oncomx_v1_0_25_small/oncomx_v1_0_25_small.sqlite\"\n",
    "print(get_database_prompt(db_path=db_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dem2sql_gpt(dem_sqls, db, db_schema):\n",
    "    prefix = \"You are a helpful assistant that translates natural language queries into SQL queries. You have access to the following database schema:\\n{db_schema}\\n\\n\\nThese are some examples of natural language queries and their corresponding SQL queries:\"\n",
    "    suffix = (\n",
    "    \"Question: {input}\\n\"\n",
    "    \"SQL:\\n\"\n",
    "    \"```sql\\n\"\n",
    "    \"-- Write only the SQL query needed to answer the question.\\n\"\n",
    "    \"-- Do not include any explanation.\\n\"\n",
    "    \"-- Always enclose your SQL in a ```sql code block.\\n\"\n",
    "    \"-- Do not add anything before or after the code block.\\n\"\n",
    ")\n",
    "    \n",
    "    examples = dem_sqls[\"demonstrations\"]\n",
    "    question = dem_sqls[\"question\"]\n",
    "    \n",
    "    example_prompt = PromptTemplate(\n",
    "        input_variables=[\"input\"],\n",
    "        template=\"Question: {input}\\nSQL: {query}\",\n",
    "        )\n",
    "    \n",
    "    prompt = FewShotPromptTemplate(\n",
    "        prefix=prefix,\n",
    "        examples=examples,\n",
    "        example_prompt=example_prompt,\n",
    "        suffix=suffix,\n",
    "        input_variables=[\"input\", \"db_schema\"],\n",
    "    )\n",
    "    \n",
    "    final_prompt = prompt.format(\n",
    "        input=question,\n",
    "        db_schema=db_schema\n",
    "        )\n",
    "    \n",
    "    response = model.invoke(final_prompt)\n",
    "    return response\n",
    "\n",
    "def extract_sql(response):\n",
    "     \n",
    "    raw_content = response.content\n",
    "\n",
    "    # Extract content between ```sql and ```\n",
    "    match = re.search(r\"```sql\\s*(.*?)\\s*```\", raw_content, re.DOTALL)\n",
    "    #match = re.search(r\"&\\s*(.*?)\\s*&\", raw_content, re.DOTALL)\n",
    "    if match:\n",
    "        sql_query = match.group(1).strip()\n",
    "    else:\n",
    "        print(raw_content)\n",
    "        sql_query = None\n",
    "        #raise ValueError(\"No SQL query found in model response.\")\n",
    "\n",
    "    return sql_query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_schema = get_database_prompt(db_path=db_path)\n",
    "for dem_sqls, file_name in zip(data, file_names):\n",
    "    with open(f\"pred_sqls_oncomx/pred_sqls_gpt4omini_{file_name[19:-5]}.txt\", \"w\") as f:\n",
    "        for dem_sql in dem_sqls:\n",
    "            response = dem2sql_gpt(dem_sql, db, db_schema)\n",
    "            generated_sql = extract_sql(response)\n",
    "\n",
    "            if generated_sql is not None:\n",
    "                # Remove the trailing newline character\n",
    "                # and add a space before writing to the file\n",
    "                f.write(generated_sql.replace(\"\\n\", \" \") + \"\\n\")\n",
    "            else: f.write(\"None\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ollama",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
